### PR TITLE
Implement user authentication for using the comment submission & history features.

### PR DESCRIPTION
**Pull request by KathyFeiyang**
### Implement user authentication for using the comment submission & history features:
- Write authentication Java servlet to obtain the current user's status, information and log-in/out URLs.
- The comment submission form no longer requires the user to fill in an email address, as the email is inferred from the log-in information; the form still requires a name, which can be a nickname.
- Add an authentication section to the home page; this section:
     1. displays the current authentication status (whether the user is currently logged in);
     2. uses a name to address the user;
     3. provides a link to log in or log out.

- The authentication section addresses the current logged-in user using the name specified by the user in the latest comment submitted by the same user (identified by a unique ID); if the user hasn't submitted any comment, the section addresses the current logged-in user with the user's email.
- The authentication section immediately switches to the specified name after the user submits a new comment.
- The comment submission and comment history sections are disabled (removed from the homepage) when the user is not logged in, to protect the comments data.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+public class AuthenticationServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    AuthenticationInfo authenticationInfo;
+    response.setContentType("application/json;");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String urlToRedirectToAfterUserLogsOut = "/index.html#contact_me";
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java]: Make /index.html#contact_me a final constant since you use it in multiple places

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -50,26 +83,48 @@ async function toggleCommentHistorySection() {
...
-  // If the comment history section was disabled just now, backup the current
-  // content and remove that content from the page.
-  if (!enableCommentHistorySection) {
+  await getAuthentication();
+
+  // If the comment history section was disabled just now or if the user is
+  // not logged in, backup the current comment history content and remove
+  // that content from the page. Additionally, if the user is not logged in,
+  // backup and remove the comment submission form.
+  if (!enableCommentHistorySection || !isUserLoggedIn) {
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/script.js:135]: Can you rewrite this if else logic? You're checking the same thing e.g. isUserLoggedIn multiple times. Perhaps think about separating out the logic into 2 helper functions: one for comment history section and one for comment submission section?

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 2 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js:135]: Got it, thank you!

Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js:135]: I separated out the logic for the comment history section and the comment submission section. I also separated out the logic for fetching comment history.
Also, unlike the comment history section (which can display different contents depending on the page number and #comments per page), the comment submission form always looks the same. So, I removed the extra logic for backing up & restoring the submission form, and changed the code to directly hide & display the submission form.

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -80,21 +79,60 @@ async function toggleCommentHistorySection() {
 }
 
 /**
- * Fetches and adds a history of comments, and the theoretical maximum and
+ * Comments include (1) comment submission; (2) comment history.
+ * * Authenticates the current user.
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/script.js]: nit. I'm kind of confused why you're adding two asterisks in some of these comments? Are you trying to tie them to (1) and (2)? If so, that should be more clear. If not, get rid of the extra asterisk

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -80,21 +79,60 @@ async function toggleCommentHistorySection() {
...
+ * attack.
  */
 async function addComments(pageId) {
+  await getAuthentication();
+
+  // Check whether the comment submission and history sections are enabled,
+  // and backup & restore the sections as needed. If the comment history
+  // section's content is fully prepared, we no longer need to fetch from
+  // the backend database.
+  checkCommentSubmissionSection();
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/script.js:104]: Why run the check twice?

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -80,21 +79,60 @@ async function toggleCommentHistorySection() {
...
+ * section, and whether the user is logged in.
+ * * Shows a warning if the user input value of the number of comments to
+ * display or page ID is invalid, or if the latest user comment may be a XSS
+ * attack.
  */
 async function addComments(pageId) {
+  await getAuthentication();
+
+  // Check whether the comment submission and history sections are enabled,
+  // and backup & restore the sections as needed. If the comment history
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/script.js]: The first sentence in this comment is kind of unnecessary given that you essentially say the same thing in the method description. In general, your code should speak for itself and that's where good variable and function names come in

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -80,21 +79,60 @@ async function toggleCommentHistorySection() {
...
+/**
+ * Checks whether the comment submission section should be enabled, and hides
+ * or displays the comment submission section as needed.
+ */
+function checkCommentSubmissionSection() {
   const commentSubmissionSection = document
       .getElementById('comment-submission-section');
+  let displayStatus;
+
+  // If the user is not logged in, hide the comment submission form. Otherwise
```
Comment (suparnabarua) [inline portfolio/src/main/webapp/script.js]: You can shorten this to : "Only display the comment submission form if the user is logged in". In fact this could be moved to the function description

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js]: Yeah, I was trying to creating a list. I'll make this clearer

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js:104]: The comment submission is disabled when the user is not logged in, but the comment history section is disabled either when the user is not logged in, or when the user chooses to disable the comment history with a checkbox.

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js]: Got it, thanks!

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js]: Will do!

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (suparnabarua) [reply portfolio/src/main/webapp/script.js:104]: oh I didn't notice it was submission vs history.  my bad!

Comment (KathyFeiyang) [reply portfolio/src/main/webapp/script.js:104]: Ah no worries at all!

**Review by suparnabarua — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    AuthenticationInfo authenticationInfo;
+    response.setContentType("application/json;");
+
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+      String urlToRedirectToAfterUserLogsOut = DataServlet.REDIRECT_URL;
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      // Obtain the name to refer to the current user.
+      String userName = this.getUserName(userService.getCurrentUser().getUserId());
+      if (!userName.isEmpty()) {
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java]: Maybe you could also send the current user email as well, so that you can display a message like "Currently logged in as kellytan@google.com" near the "Hello Kelly", since there isn't necessarily any other indication of which email you're logging in as if you're already logged in on another page.

**Review by kellyktan — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java (hunk 7 above)
Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java]: Done! Should I complete the merge?

Comment (kellyktan) [reply portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java]: Yes, looks good!

Comment (KathyFeiyang) [reply portfolio/src/main/java/com/google/sps/servlets/AuthenticationServlet.java]: Thanks!
